custom_components/bosch_alarm/alarm_control_panel.py

```diff
@@ -52,35 +67,48 @@ def supported_features(self) -> int:
             AlarmControlPanelEntityFeature.ARM_HOME
             | AlarmControlPanelEntityFeature.ARM_AWAY
         )
-
-    async def async_alarm_disarm(self, code=None) -> None:
-        await self._panel.area_disarm(self._area_id)
-    async def async_alarm_arm_home(self, code=None) -> None:
-        await self._panel.area_arm_part(self._area_id)
-    async def async_alarm_arm_away(self, code=None) -> None:
-        await self._panel.area_arm_all(self._area_id)
+    
+    def _arming_code_correct(self, code) -> bool:
+        return not self._arming_code or int(code) == self._arming_code
+
+    async def async_alarm_disarm(self, code=0) -> None:
+        if self._arming_code_correct(code): 
+            await self._panel.area_disarm(self._area_id)
+    async def async_alarm_arm_home(self, code=0) -> None:
+        if self._arming_code_correct(code): 
+            await self._panel.area_arm_part(self._area_id)
+    async def async_alarm_arm_away(self, code=0) -> None:
+        if self._arming_code_correct(code): 
+            await self._panel.area_arm_all(self._area_id)
 
     @property
     def extra_state_attributes(self):
         ready_state = READY_STATE_NO
         if self._area.all_ready: ready_state = READY_STATE_AWAY
         elif self._area.part_ready: ready_state = READY_STATE_HOME
         return { READY_STATE_ATTR: ready_state,
-                 FAULTED_POINTS_ATTR: self._area.faults }
+                 FAULTED_POINTS_ATTR: self._area.faults,
+                 ALARMS_ATTR: "\n".join(self._area.alarms) }
 
     async def async_added_to_hass(self):
         self._area.status_observer.attach(self.async_schedule_update_ha_state)
+        self._area.alarm_observer.attach(self.async_schedule_update_ha_state)
         self._area.ready_observer.attach(self.async_schedule_update_ha_state)
 
     async def async_will_remove_from_hass(self):
         self._area.status_observer.detach(self.async_schedule_update_ha_state)
+        self._area.alarm_observer.detach(self.async_schedule_update_ha_state)
         self._area.ready_observer.detach(self.async_schedule_update_ha_state)
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
     """Set up control panels for each area."""
 
     panel = hass.data[DOMAIN][config_entry.entry_id]
+
+    arming_code = None
+    if CONF_CODE in config_entry.options:
```

Comment:
Is options a dict? Then you should be able to write just

arming_code = config_entry.options.get(CONF_CODE, None)